File: acorn-core/src/main/java/com/ardley/acorn/permission/PermissionSet.java
package com.ardley.acorn.permission;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * A role's complete permission configuration defining which actions are allowed or denied
 * and under what scope constraints.
 *
 * <p>The permission set is the central data structure of Acorn's authorization model.
 * It is loaded from persistent storage via the
 * {@link com.ardley.acorn.store.PermissionStore} and evaluated by the
 * {@link com.ardley.acorn.evaluator.Evaluator}.
 *
 * <p>JSON format:
 * <pre>{@code
 * {
 *   "allow": {
 *     "ListUsers": "all",
 *     "UpdateUser": {"department": {"match": "principal"}}
 *   },
 *   "deny": {
 *     "DeleteUser": "all"
 *   }
 * }
 * }</pre>
 *
 * <p>Superadmin shorthand: {@code {"allow": "all"}}
 */
public final class PermissionSet {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final boolean allowAll;
    private final Map<String, PermissionLevel> allow;
    private final Map<String, PermissionLevel> deny;

    private PermissionSet(boolean allowAll, Map<String, PermissionLevel> allow, Map<String, PermissionLevel> deny) {
        this.allowAll = allowAll;
        this.allow = Map.copyOf(allow);
        this.deny = Map.copyOf(deny);
    }

    public static PermissionSet allowAll() {
        return new PermissionSet(true, Map.of(), Map.of());
    }

    public static PermissionSet empty() {
        return new PermissionSet(false, Map.of(), Map.of());
    }

    public boolean isAllowAll() {
        return allowAll;
    }

    public boolean hasAllowFor(String actionName) {
        if (allowAll) return true;
        PermissionLevel level = allow.get(actionName);
        return level != null && !level.isNone();
    }

    public boolean hasUnconditionalDeny(String actionName) {
        PermissionLevel level = deny.get(actionName);
        return level != null && level.isAll();
    }

    public Optional<PermissionLevel> allowLevel(String actionName) {
        return Optional.ofNullable(allow.get(actionName));
    }

    public Optional<PermissionLevel> denyLevel(String actionName) {
        return Optional.ofNullable(deny.get(actionName));
    }

    /**
     * Parses a permission set from a JSON string.
     * Supports the {@code {"allow": "all"}} shorthand for superadmin roles.
     *
     * @param json the JSON string to parse
     * @return the parsed permission set
     * @throws IllegalArgumentException if the JSON is malformed
     */
    public static PermissionSet fromJson(String json) {
        try {
            JsonNode root = MAPPER.readTree(json);
            return fromJsonNode(root);
        } catch (IllegalArgumentException e) {
            throw e;
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid permission JSON: " + e.getMessage(), e);
        }
    }

    public static PermissionSet fromJsonNode(JsonNode root) {
        JsonNode allowNode = root.get("allow");
        JsonNode denyNode = root.get("deny");

        if (allowNode != null && allowNode.isTextual() && "all".equals(allowNode.asText())) {
            return new PermissionSet(true, Map.of(), parseLevelMap(denyNode));
        }

        return new PermissionSet(false, parseLevelMap(allowNode), parseLevelMap(denyNode));
    }

    private static Map<String, PermissionLevel> parseLevelMap(JsonNode node) {
        if (node == null || node.isNull()) {
            return Map.of();
        }
        Map<String, PermissionLevel> map = new HashMap<>();
        var fields = node.fields();
        while (fields.hasNext()) {
            var entry = fields.next();
            map.put(entry.getKey(), PermissionLevel.fromJson(entry.getValue()));
        }
        return map;
    }
}

File: acorn-node/packages/core/src/permission-set.ts
import type { PermissionLevel } from './permission-level.js';
import { parsePermissionLevel } from './permission-level.js';

export class PermissionSet {
  private constructor(
    private readonly _allowAll: boolean,
    private readonly allow: ReadonlyMap<string, PermissionLevel>,
    private readonly deny: ReadonlyMap<string, PermissionLevel>,
  ) {}

  static allowAll(): PermissionSet {
    return new PermissionSet(true, new Map(), new Map());
  }

  static empty(): PermissionSet {
    return new PermissionSet(false, new Map(), new Map());
  }

  static fromJson(json: string): PermissionSet {
    try {
      const root = JSON.parse(json);
      return PermissionSet.fromObject(root);
    } catch (e) {
      if (e instanceof Error && e.message.startsWith('Invalid permission')) throw e;
      throw new Error(`Invalid permission JSON: ${(e as Error).message}`);
    }
  }

  static fromObject(root: unknown): PermissionSet {
    if (root === null || typeof root !== 'object') {
      throw new Error('Permission set must be an object');
    }
    const obj = root as Record<string, unknown>;
    const allowNode = obj.allow;
    const denyNode = obj.deny;

    if (typeof allowNode === 'string' && allowNode === 'all') {
      return new PermissionSet(true, new Map(), parseLevelMap(denyNode));
    }

    return new PermissionSet(false, parseLevelMap(allowNode), parseLevelMap(denyNode));
  }

  isAllowAll(): boolean {
    return this._allowAll;
  }

  hasAllowFor(actionName: string): boolean {
    if (this._allowAll) return true;
    const level = this.allow.get(actionName);
    return level !== undefined && level.type !== 'none';
  }

  hasUnconditionalDeny(actionName: string): boolean {
    const level = this.deny.get(actionName);
    return level !== undefined && level.type === 'all';
  }

  allowLevel(actionName: string): PermissionLevel | undefined {
    return this.allow.get(actionName);
  }

  denyLevel(actionName: string): PermissionLevel | undefined {
    return this.deny.get(actionName);
  }
}

function parseLevelMap(node: unknown): Map<string, PermissionLevel> {
  if (node === null || node === undefined) return new Map();
  if (typeof node !== 'object') throw new Error('Level map must be an object');
  const map = new Map<string, PermissionLevel>();
  for (const [key, value] of Object.entries(node as Record<string, unknown>)) {
    map.set(key, parsePermissionLevel(value));
  }
  return map;
}
